File: 07-02-2024_MIn_Distance_between_two_given_nodes _of_a_binary_tree.java
class GfG {
    Node findNearestCommonRoot(Node root, int a, int b) {
        if (root == null || root.data == a || root.data == b)
            return root;

        Node l = findNearestCommonRoot(root.left, a, b);
        Node r = findNearestCommonRoot(root.right, a, b);

        if (l == null)
            return r;
        else if (r == null)
            return l;
        else
            return root;
    }

    int findDistance(Node root, int n) {
        if (root == null)
            return 0;

        if (root.data == n)
            return 1;

        int p = findDistance(root.left, n);
        int q = findDistance(root.right, n);

        if (p == 0 && q == 0)
            return 0;

        return p + q + 1;
    }

    int findDist(Node root, int a, int b) {
        Node LCA = findNearestCommonRoot(root, a, b);

        return findDistance(LCA, a) + findDistance(LCA, b) - 2;
    }
}

File: 07-02-2024_MIn_Distance_between_two_given_nodes _of_a_binary_tree.cpp

class Solution{
    public:
    
    Node* find_nearest_common_root(Node* root , int a , int b){
        if(root==NULL||root->data==a||root->data==b)
        return root ;
        
        Node* l = find_nearest_common_root(root->left , a, b);
        Node* r = find_nearest_common_root(root->right, a, b);
        
        if(l==NULL)
        return r;
        else if(r==NULL)
        return l;
        else 
        return root ;
    }
    
    int find_dis(Node* root , int n){
        if(root==NULL) return 0;
        
        if(root->data==n) return 1;
        
        int p = find_dis(root->left,n);
        int q = find_dis(root->right,n);
        
        if(p==0 and q==0) return 0;
        
        return p+q+1;
    }
    
    /* Should return minimum distance between a and b
    in a tree with given root*/
    int findDist(Node* root, int a, int b) {
        // Your code here
        
        Node* LCA = find_nearest_common_root(root,a,b);
        
        return find_dis(LCA,a)+find_dis(LCA,b) -2;
    }
};
